File: Ocean.studio-main/electron/services/agent.ts
import { EventEmitter } from 'events';
import { spawn, ChildProcess } from 'child_process';
import path from 'path';
import { fileURLToPath } from 'url';
import type { TerminalManager } from './terminal.js';
import type { TerminalType } from './terminal.js';
import type { FileSystemService } from './filesystem.js';
import type { PortPreviewService } from './port-preview.js';
import type { McpManager } from './mcp-manager.js';
import type { GitHubService } from './github.js';
import { resolvePreferredTerminal, type PlatformContext } from './terminal-router.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

export type AgentMode = 'review' | 'auto' | 'bypass';

export interface AgentEvent {
  type: 'thought' | 'action' | 'command' | 'todo' | 'result' | 'error' | 'status';
  id: string;
  timestamp: number;
  summary: string;
  detail?: string;
  duration?: number;
  status?: 'running' | 'completed' | 'failed' | 'paused';
  metadata?: Record<string, unknown>;
}

export class AgentService extends EventEmitter {
  private status: 'idle' | 'running' | 'paused' = 'idle';
  private currentProcess: ChildProcess | null = null;
  private mode: AgentMode = 'review';
  private processingQueue = false;
  private messageQueue: { message: string; context: unknown; resolve: (text: string) => void; reject: (e: Error) => void }[] = [];
  private platformContext: PlatformContext = {
    platform: process.platform,
    isElectron: true,
    isDesktop: true,
    isMobile: false,
    isWeb: false,
    hasNativeShell: true,
    shellLabel: process.platform === 'win32' ? 'PowerShell' : 'Bash',
  };

  constructor(
    private terminal: TerminalManager,
    private fs: FileSystemService,
    private preview: PortPreviewService,
    private mcp?: McpManager,
    private github?: GitHubService
  ) {
    super();
  }

  setPlatformContext(ctx: PlatformContext) {
    this.platformContext = ctx;
  }

  setMode(mode: AgentMode) {
    this.mode = mode;
  }

  getStatus() {
    return this.status;
  }

  getPreferredTerminalType(): TerminalType {
    return resolvePreferredTerminal(this.platformContext);
  }

  pause() {
    if (this.currentProcess) {
      this.currentProcess.kill('SIGTERM');
      this.currentProcess = null;
    }
    this.messageQueue = [];
    this.processingQueue = false;
    this.status = 'paused';
    this.emitEvent({
      type: 'status',
      id: crypto.randomUUID(),
      timestamp: Date.now(),
      summary: 'Agent session paused',
      status: 'paused',
    });
  }

  async sendMessage(message: string, context: unknown): Promise<string> {
    return new Promise((resolve, reject) => {
      this.messageQueue.push({ message, context, resolve, reject });
      void this.processQueue();
    });
  }

  private async processQueue(): Promise<void> {
    if (this.processingQueue) return;
    this.processingQueue = true;

    while (this.messageQueue.length > 0) {
      const item = this.messageQueue.shift()!;
      try {
        const response = await this.executeSend(item.message, item.context);
        item.resolve(response);
      } catch (err) {
        item.reject(err instanceof Error ? err : new Error('Agent send failed'));
      }
    }

    this.processingQueue = false;
  }

  private async executeSend(message: string, context: unknown): Promise<string> {
    if (this.status === 'running') {
      // Wait briefly for prior send to finish (queue should prevent this)
      await this.delay(100);
    }
    this.status = 'running';

    const ctx = context as Record<string, unknown>;
    const phase = ctx?._multiAgentPhase as string | undefined;
    const isSubAgentPass = phase === 'parallel-worker' || phase === 'collaborative';
    const memberCtx = ctx?.multiAgentMember as { memberName?: string; role?: string } | undefined;

    const agentMode = (ctx?.agentMode as AgentMode) ?? this.mode;
    this.mode = agentMode;

    const terminalType = (ctx?.terminalType as TerminalType) || this.getPreferredTerminalType();
    const workspacePath = ctx?.workspacePath as string | undefined;
    const mcpFromCtx = ctx?.mcpServers as Record<string, unknown> | undefined;
    const mcpConfig = mcpFromCtx && Object.keys(mcpFromCtx).length > 0
      ? { mcpServers: mcpFromCtx, connectedCount: (ctx?.mcpMeta as { connectedCount?: number })?.connectedCount }
      : this.mcp ? await this.mcp.getAgentConfig() : {};
    const ghStatus = this.github?.getStatus();

    const thoughtId = crypto.randomUUID();
    const startTime = Date.now();

    if (!isSubAgentPass) {
      this.emitEvent({
        type: 'thought',
        id: thoughtId,
        timestamp: startTime,
        summary: 'Planning next moves',
        status: 'running',
      });
      await this.delay(600);
    } else {
      await this.delay(150);
    }

    this.emitEvent({
      type: 'thought',
      id: thoughtId,
      timestamp: Date.now(),
      summary: isSubAgentPass
        ? `${memberCtx?.memberName ?? 'Sub-agent'} (${phase})`
        : `Thought for ${Math.max(1, Math.round((Date.now() - startTime) / 1000))}s`,
      detail: this.generateThoughtDetail(message, ctx, terminalType),
      duration: Date.now() - startTime,
      status: 'completed',
      metadata: isSubAgentPass ? { memberId: memberCtx, phase } : undefined,
    });

    const actionId = crypto.randomUUID();
    this.emitEvent({
      type: 'action',
      id: actionId,
      timestamp: Date.now(),
      summary: isSubAgentPass
        ? `${memberCtx?.memberName ?? 'Agent'}: ${message.slice(0, 60)}...`
        : `Processing: ${message.slice(0, 80)}${message.length > 80 ? '...' : ''}`,
      status: 'running',
      metadata: { terminalType, phase },
    });

    const modelConfig = ctx?.modelConfig as { agentTimeoutSec?: number; antiTimeout?: boolean } | undefined;
    const antiTimeout = modelConfig?.antiTimeout ?? false;
    const timeoutSec = antiTimeout ? 0 : (modelConfig?.agentTimeoutSec ?? 120);

    try {
      if (!isSubAgentPass) {
        const agentTermId = this.terminal.ensureAgentTerminal(terminalType, workspacePath);
        this.emitEvent({
          type: 'command',
          id: crypto.randomUUID(),
          timestamp: Date.now(),
          summary: `Agent terminal ready (${terminalType})`,
          detail: `Session: ${agentTermId}\nPlatform: ${this.platformContext.platform}`,
          status: 'completed',
        });
      }

      await this.runAgentScript(message, {
        ...ctx,
        terminalType,
        agentTerminalId: 'ocean-agent-terminal',
        mcpServers: mcpConfig,
        github: ghStatus,
      }, { timeoutSec, antiTimeout });

      const responseText = this.lastScriptResponse || message.slice(0, 500);

      // Sub-agent passes don't auto-run shell commands — only head/user messages do
      if (!isSubAgentPass && this.looksLikeCommand(message)) {
        const safe = this.checkCommandSafety(message);
        const agentMode = (ctx?.agentMode as AgentMode) ?? this.mode;
        const effectiveMode = agentMode === 'bypass' ? 'bypass' : agentMode;

        if (safe.requiresApproval && effectiveMode === 'review') {
          this.emitEvent({
            type: 'command',
            id: crypto.randomUUID(),
            timestamp: Date.now(),
            summary: `Awaiting approval: ${message.slice(0, 60)}`,
            detail: safe.reason,
            status: 'running',
            metadata: { requiresApproval: true, command: message },
          });
        } else {
          this.terminal.runAgentCommand(message, terminalType, workspacePath);
          this.emitEvent({
            type: 'command',
            id: crypto.randomUUID(),
            timestamp: Date.now(),
            summary: `Ran command in ${terminalType} terminal`,
            detail: `$ ${message}`,
            status: 'completed',
          });
        }
      }

      this.emitEvent({
        type: 'result',
        id: actionId,
        timestamp: Date.now(),
        summary: responseText.slice(0, 120) + (responseText.length > 120 ? '...' : ''),
        detail: responseText,
        status: 'completed',
        metadata: { responseText },
      });

      return responseText;
    } catch (err) {
      this.emitEvent({
        type: 'error',
        id: actionId,
        timestamp: Date.now(),
        summary: `Error: ${err instanceof Error ? err.message : 'Unknown error'}`,
        status: 'failed',
      });
      throw err instanceof Error ? err : new Error('Agent execution failed');
    } finally {
      this.status = 'idle';
    }
  }

  private lastScriptResponse = '';

  private looksLikeCommand(message: string): boolean {
    const trimmed = message.trim();
    // Any line starting with $ or common shell patterns is a command
    if (/^\$/.test(trimmed)) return true;
    // Broad match — terminals support ALL commands, not a whitelist
    const shellPattern = /^[\w./~-]+(\s+|$)/;
    return shellPattern.test(trimmed) && trimmed.length < 500;
  }

  private checkCommandSafety(command: string): { requiresApproval: boolean; reason: string } {
    // Only flag truly catastrophic operations — everything else runs freely
    const catastrophic = /rm\s+(-[a-zA-Z]*f[a-zA-Z]*\s+.*\/|.*\s+\/(\s|$))|sudo\s+rm\s+|mkfs\.|dd\s+if=.*of=\/dev\/|:\(\)\{.*\};/i;
    if (catastrophic.test(command)) {
      return { requiresApproval: true, reason: 'Potentially destructive system command' };
    }
    return { requiresApproval: false, reason: '' };
  }

  private async runAgentScript(
    message: string,
    context: Record<string, unknown>,
    opts: { timeoutSec: number; antiTimeout: boolean } = { timeoutSec: 120, antiTimeout: false }
  ) {
    const scriptPath = path.join(__dirname, '../../agent/agent.py');
    const mode = (context.agentMode as AgentMode) ?? this.mode;
    const payload = JSON.stringify({
      message,
      context: {
        ...context,
        agentTimeoutSec: opts.timeoutSec,
        antiTimeout: opts.antiTimeout,
      },
      mode,
    });
    this.lastScriptResponse = '';

    return new Promise<void>((resolve, reject) => {
      this.currentProcess = spawn('python3', [scriptPath], {
        env: { ...process.env, OCEAN_AGENT_PAYLOAD: payload },
      });

      let stdout = '';
      let stderr = '';
      let timeoutId: ReturnType<typeof setTimeout> | null = null;
      let extensionId: ReturnType<typeof setInterval> | null = null;

      const clearTimers = () => {
        if (timeoutId) clearTimeout(timeoutId);
        if (extensionId) clearInterval(extensionId);
        timeoutId = null;
        extensionId = null;
      };

      if (opts.antiTimeout) {
        extensionId = setInterval(() => {
          this.emitEvent({
            type: 'status',
            id: crypto.randomUUID(),
            timestamp: Date.now(),
            summary: 'Anti-timeout: extending session until workflow completes',
            status: 'running',
          });
        }, 600_000);
      } else if (opts.timeoutSec > 0) {
        timeoutId = setTimeout(() => {
          if (this.currentProcess) {
            this.currentProcess.kill('SIGTERM');
            this.currentProcess = null;
          }
          clearTimers();
          reject(new Error(`Agent timeout after ${opts.timeoutSec}s`));
        }, opts.timeoutSec * 1000);
      }

      this.currentProcess.stdout?.on('data', (data: Buffer) => {
        stdout += data.toString();
        const lines = stdout.split('\n');
        stdout = lines.pop() || '';
        for (const line of lines) {
          if (line.startsWith('OCEAN_EVENT:')) {
            try {
              this.emitEvent(JSON.parse(line.slice(12)));
            } catch { /* ignore */ }
          } else if (line.startsWith('OCEAN_RESPONSE:')) {
            try {
              const parsed = JSON.parse(line.slice(15));
              this.lastScriptResponse = parsed.content ?? String(parsed);
            } catch {
              this.lastScriptResponse = line.slice(15);
            }
          }
        }
      });

      this.currentProcess.stderr?.on('data', (data: Buffer) => {
        stderr += data.toString();
      });

      this.currentProcess.on('close', (code) => {
        clearTimers();
        this.currentProcess = null;
        if (code === 0) resolve();
        else reject(new Error(stderr || `Agent exited with code ${code}`));
      });
    });
  }

  private generateThoughtDetail(message: string, ctx: Record<string, unknown> | null, terminalType: TerminalType): string {
    const parts = [`User request: ${message}`];
    if (ctx?.workspacePath) parts.push(`Workspace: ${ctx.workspacePath}`);
    if (ctx?.activeFile) parts.push(`Active file: ${ctx.activeFile}`);
    if (ctx?.agentMode) parts.push(`Mode: ${ctx.agentMode}`);
    const modelConfig = ctx?.modelConfig as Record<string, unknown> | undefined;
    if (modelConfig?.modelId) parts.push(`Model: ${modelConfig.modelId} (temp=${modelConfig.temperature}, thinking=${modelConfig.thinkingLevel})`);
    else if (modelConfig?.selectedModelId) parts.push(`Model: ${modelConfig.selectedModelId}`);
    const mcpMeta = ctx?.mcpMeta as { connectedCount?: number } | undefined;
    if (mcpMeta?.connectedCount) parts.push(`MCP connectors: ${mcpMeta.connectedCount}`);
    const pluginTools = ctx?.pluginTools as unknown[] | undefined;
    if (pluginTools?.length) parts.push(`Plugin tools: ${pluginTools.length}`);
    const providerMeta = ctx?.providerMeta as { connectedCount?: number; activeProviderIds?: string[] } | undefined;
    if (providerMeta?.connectedCount) {
      parts.push(`Providers connected: ${providerMeta.connectedCount}`);
      if (providerMeta.activeProviderIds?.length) {
        parts.push(`Active in chat: ${providerMeta.activeProviderIds.join(', ')}`);
      }
    }
    const providerCtx = ctx?.providerContext as { proxyStatus?: { running: boolean; url: string } } | undefined;
    if (providerCtx?.proxyStatus?.running) {
      parts.push(`Ocean proxy: ${providerCtx.proxyStatus.url}`);
    }
    const multiAgent = ctx?.multiAgent as { enabled?: boolean; comboName?: string; mode?: string; members?: unknown[] } | undefined;
    if (multiAgent?.enabled) {
      parts.push(`Multi-agent team: ${multiAgent.comboName} (${multiAgent.mode})`);
      parts.push(`Team size: ${(multiAgent.members as unknown[])?.length ?? 0} agents`);
    }
    const member = ctx?.multiAgentMember as { memberName?: string; role?: string; providerId?: string; modelId?: string } | undefined;
    if (member?.memberName) {
      parts.push(`Agent: ${member.memberName} (${member.role}) — ${member.providerId}/${member.modelId}`);
    }
    const phase = ctx?._multiAgentPhase as string | undefined;
    if (phase) parts.push(`Workflow phase: ${phase}`);
    parts.push(`Terminal backend: ${terminalType}`);
    parts.push(`Platform: ${this.platformContext.platform} (electron=${this.platformContext.isElectron})`);
    if (terminalType === 'shell') {
      parts.push('Using real host shell — all installed binaries available.');
    } else if (terminalType === 'cloud') {
      parts.push('Using Google Cloud Shell — Debian VM with apt, pip, gcloud.');
    } else {
      parts.push('Using native mobile Linux environment.');
    }
    if (this.mode === 'review') parts.push('Review mode: only catastrophic commands need approval.');
    else parts.push('All terminal commands run without restriction.');
    return parts.join('\n');
  }

  private emitEvent(event: AgentEvent) {
    this.emit('event', event);
  }

  private delay(ms: number) {
    return new Promise((r) => setTimeout(r, ms));
  }
}

File: agent/agent.py
#!/usr/bin/env python3
"""Ocean.studio Agent — hardware-level coding agent backend."""

import json
import os
import sys
import time
import uuid
from pathlib import Path


def emit_event(event_type: str, summary: str, detail: str = "", status: str = "completed", metadata: dict | None = None):
    event = {
        "type": event_type,
        "id": str(uuid.uuid4()),
        "timestamp": int(time.time() * 1000),
        "summary": summary,
        "detail": detail,
        "status": status,
        "metadata": metadata or {},
    }
    print(f"OCEAN_EVENT:{json.dumps(event)}", flush=True)


def resolve_terminal_guidance(terminal_type: str) -> str:
    guides = {
        "shell": (
            "SHELL terminal — real host OS shell via node-pty.\n"
            "- All commands run without whitelist restriction\n"
            "- Windows: PowerShell (WSL/bash if WSL installed)\n"
            "- Linux/Mac: Bash/Zsh with apt/brew/pip and all host binaries"
        ),
        "cloud": (
            "CLOUD terminal — Google Cloud Shell (Debian VM in GCP).\n"
            "- OAuth grants user-level GCP access\n"
            "- apt, pip, gcloud, docker, persistent $HOME"
        ),
        "native": (
            "NATIVE terminal — mobile Linux via proot (APK builds).\n"
            "- Full package manager without root\n"
            "- All Linux commands available in proot environment"
        ),
    }
    return guides.get(terminal_type, f"Unknown terminal type: {terminal_type}")


def format_model_context(model_config: dict) -> str:
    if not model_config:
        return "Model: default"
    lines = [
        f"Model: {model_config.get('modelId', 'unknown')}",
        f"Temperature: {model_config.get('temperature', 0.7)}",
        f"Thinking: {model_config.get('thinkingLevel', 'medium')}",
        f"Strictness: {model_config.get('strictness', 'normal')}",
        f"Scraping: {model_config.get('scrapingProvider', 'firecrawl')}",
    ]
    if model_config.get("bypassDefaultInstructions"):
        lines.append("Bypass default instructions: yes")
    if model_config.get("customInstructions"):
        lines.append(f"Custom instructions: {str(model_config['customInstructions'])[:200]}")
    skills = model_config.get("skills") or []
    if skills:
        lines.append(f"Skills loaded: {len(skills)}")
    webhooks = model_config.get("webhooks") or []
    if webhooks:
        lines.append(f"Webhooks: {len(webhooks)}")
    custom_fn = model_config.get("customFunctions") or []
    if custom_fn:
        lines.append(f"Custom functions: {len(custom_fn)}")
    if model_config.get("postgresConnectionString"):
        lines.append("PostgreSQL: configured")
    return "\n".join(lines)


def is_catastrophic_command(command: str) -> bool:
    destructive = ["rm -rf /", "sudo rm", "mkfs.", "dd if=", "of=/dev/"]
    lower = command.lower()
    return any(d in lower for d in destructive)


def main():
    payload_str = os.environ.get("OCEAN_AGENT_PAYLOAD", "{}")
    payload = json.loads(payload_str)
    message = payload.get("message", "")
    context = payload.get("context", {})
    mode = payload.get("mode", "review")

    workspace = context.get("workspacePath", "")
    active_file = context.get("activeFile", "")
    terminal_type = context.get("terminalType", "shell")
    agent_terminal_id = context.get("agentTerminalId", "ocean-agent-terminal")
    model_config = context.get("modelConfig") or {}
    mcp_servers = context.get("mcpServers") or {}
    mcp_meta = context.get("mcpMeta") or {}
    plugin_tools = context.get("pluginTools") or []
    capabilities = context.get("capabilities") or {}
    platform_info = context.get("platform") or {}
    provider_meta = context.get("providerMeta") or {}
    provider_context = context.get("providerContext") or {}
    providers = context.get("providers") or {}
    multi_agent = context.get("multiAgent") or {}

    thought_detail = "\n".join([
        f"Message: {message}",
        f"Terminal: {terminal_type}",
        f"Mode: {mode}",
        f"Platform: {platform_info.get('platform', 'unknown')}",
        format_model_context(model_config),
        f"MCP connectors: {mcp_meta.get('connectedCount', len(mcp_servers))}",
        f"Plugin tools: {len(plugin_tools)}",
        f"Providers connected: {provider_meta.get('connectedCount', len(providers))}",
        f"Active providers: {', '.join(provider_meta.get('activeProviderIds') or []) or 'none'}",
        f"Multi-agent: {multi_agent.get('comboName', 'off')} ({multi_agent.get('mode', 'single')})" if multi_agent.get('enabled') else "Multi-agent: disabled",
        "",
        resolve_terminal_guidance(terminal_type),
    ])

    if provider_context.get("serialized"):
        emit_event(
            "action",
            "Provider agent context loaded",
            str(provider_context.get("serialized", ""))[:500],
            "completed",
            {"activeProviders": provider_meta.get("activeProviderIds", [])},
        )

    if multi_agent.get("enabled"):
        emit_event(
            "action",
            f"Multi-agent team: {multi_agent.get('comboName', 'team')}",
            str(multi_agent.get("serializedCombo", ""))[:600],
            "running",
            {"mode": multi_agent.get("mode"), "memberCount": len(multi_agent.get("members") or [])},
        )

    member_ctx = context.get("multiAgentMember")
    phase = context.get("_multiAgentPhase")
    if member_ctx:
        emit_event(
            "action",
            f"Sub-agent: {member_ctx.get('memberName', 'agent')}",
            f"Role: {member_ctx.get('role')}\nProvider: {member_ctx.get('providerId')}\nModel: {member_ctx.get('modelId')}\nPhase: {phase}",
            "running",
            {"memberId": member_ctx.get("memberId"), "phase": phase},
        )

    emit_event("thought", "Analyzing request", thought_detail, "completed")

    if mcp_servers:
        emit_event(
            "action",
            f"Using {len(mcp_servers)} MCP connector(s)",
            ", ".join(mcp_servers.keys()),
            "completed",
            {"mcpCount": len(mcp_servers)},
        )

    if plugin_tools:
        emit_event(
            "action",
            f"Loaded {len(plugin_tools)} plugin tool(s)",
            ", ".join(t.get("name", "?") for t in plugin_tools[:5]),
            "completed",
        )

    skills = model_config.get("skills") or []
    if skills:
        emit_event(
            "action",
            f"Applied {len(skills)} skill file(s)",
            ", ".join(s.get("name", "?") for s in skills[:5]),
            "completed",
        )

    trimmed = message.strip().lstrip("$").strip()
    looks_like_command = bool(trimmed) and len(trimmed) < 500 and not trimmed.startswith("{")
    command_keywords = ["run", "install", "npm", "pip", "apt", "gcloud", "git", "docker", "curl", "wget"]

    if looks_like_command and (any(kw in trimmed.lower() for kw in command_keywords) or " " in trimmed):
        emit_event("action", "Preparing shell command", f"Target terminal: {terminal_type}", "running")
        time.sleep(0.2)

        requires_approval = mode == "review" and is_catastrophic_command(trimmed)

        emit_event(
            "command",
            f"Execute in {terminal_type} terminal",
            f"$ {trimmed[:200]}\nSession: {agent_terminal_id}\nRequires approval: {requires_approval}",
            "completed" if not requires_approval else "running",
            {"requiresApproval": requires_approval, "terminalType": terminal_type},
        )

    elif "preview" in message.lower() or "port" in message.lower():
        emit_event("action", "Checking port status", "Scanning terminal output for active ports", "running")
        time.sleep(0.2)
        emit_event("result", "Port scan complete", "Ports auto-register in Preview panel from terminal output", "completed")

    elif workspace:
        ws_path = Path(workspace)
        files = list(ws_path.rglob("*")) if ws_path.exists() else []
        emit_event(
            "command",
            f"Scanned workspace ({len(files)} items)",
            f"Workspace: {workspace}\nActive file: {active_file or 'none'}\nTerminal: {terminal_type}",
            "completed",
        )
    else:
        caps_platform = capabilities.get("platform", "electron")
        emit_event(
            "result",
            "Ready to assist",
            f"Terminal: {terminal_type}\nPlatform: {caps_platform}\n"
            f"Scraping providers: {', '.join(capabilities.get('scrapingProviders', []))}\n"
            "I can edit files, run commands, use MCP tools, and manage previews.",
            "completed",
        )

    emit_event("status", "Agent idle", "Waiting for next instruction", "completed")

    # Emit substantive response for orchestrator capture
    member_ctx = context.get("multiAgentMember") or {}
    response_lines = [f"## Response\n"]
    if member_ctx:
        response_lines.append(f"**{member_ctx.get('memberName', 'Agent')}** ({member_ctx.get('role', 'worker')})")
        if member_ctx.get("task"):
            response_lines.append(f"Task: {member_ctx.get('task')}")
    response_lines.append(f"\nRequest: {message[:600]}")
    if workspace and Path(workspace).exists():
        py_files = list(Path(workspace).rglob("*.py"))[:5]
        ts_files = list(Path(workspace).rglob("*.ts"))[:5]
        response_lines.append(f"\nWorkspace scan: {len(py_files)} Python, {len(ts_files)} TypeScript files nearby")
    if mcp_servers:
        response_lines.append(f"MCP active: {', '.join(list(mcp_servers.keys())[:5])}")
    if plugin_tools:
        response_lines.append(f"Plugins: {len(plugin_tools)} tools loaded")
    response_lines.append(f"\nTerminal ({terminal_type}): ready for command execution")
    if phase:
        response_lines.append(f"Workflow phase: {phase}")

    response_content = "\n".join(response_lines)
    print(f"OCEAN_RESPONSE:{json.dumps({'content': response_content})}", flush=True)


if __name__ == "__main__":
    main()
